File: plugin/g_trans.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""This script allow users to translate a string
from one language to another with Google translate"""

import codecs
import locale
import sys
import re
import urllib
import urllib2
import json

def print_params(data):
  """print parameters from list"""
  for val in data:
    if isinstance(val, basestring):
      print "\t " + val

def main():
  """
  Usage:
    first arg - string to translate
    second arg - source lang
    third arg - target lang  
  Example:
    translate.py 'text to translate' en ru
    translate.py 'text to translate' ru en
  """
  
  url = "http://translate.google.com/translate_a/t?%s"
  list_of_params = {'client' : 't', 
                    'hl' : 'en',
                    'multires' : '1', }  
  
  #all arguments given
  if len(sys.argv) == 2:
    list_of_params.update({'text' : sys.argv[1],
                           'sl' : "en",
                           'tl' : "ru"})

    request = urllib2.Request(url % urllib.urlencode(list_of_params),
      headers={ 'User-Agent': 'Mozilla/5.0', 'Accept-Charset': 'utf-8' })
    res = urllib2.urlopen(request).read()

    fixed_json = re.sub(r',{2,}', ',', res).replace(',]', ']')  
    data = json.loads(fixed_json)
#    data = data.encode('utf-8')
    
    #simple translation
    sys.stdout = codecs.getwriter(locale.getpreferredencoding())(sys.stdout) 
    print "%s / %s / %s" % (data[0][0][0], data[0][0][1],
                data[0][0][2] or data[0][0][3])
    
    #abbreviation
    if not isinstance(data[1], basestring):
      print data[1][0][0]
      print_params(data[1][0][1])
      
    #interjection  
    try:
      if not isinstance(data[1][1], basestring):
        print data[1][1][0]
        print_params(data[1][1][1])
    except Exception:
      print "no interjection"
  else:
    print main.__doc__
    
if __name__ == '__main__':
  main()

File: plugin/g_trans.rb
# translate.rb
#!/usr/local/bin/ruby
 
# install json
# sudo gem install json
 
$KCODE = 'utf'
require 'pp'
require 'net/http'
require 'uri'
require 'rubygems'
require 'json'
require 'cgi'
 
def translate( string, to, from='en' )
 
  url =
"http://translate.google.ru/translate_a/t?client=x&text={#{string}}&sl={#{from.to_s}}&tl={#{to.to_s}}"
 
  resp = Net::HTTP.get_response( URI.parse( "#{url}" ) )
 
  json = JSON.parse( resp.body )
 
  if json['responseStatus'] == 200
    json['responseData']['translatedText']
  else
    raise StandardError, resp['responseDetails']
  end
end
 
if __FILE__ == $0
  string = ARGV[0] || 'oracle'
  lang = ARGV[1] || 'ru' # Change it to your language.
  puts translate( string, lang  )
end

